### ISSUE TITLE
[FEATURE] Replay Buffer Settings

### ISSUE THREAD

**Issue by tivomk** [AdamRamberg/unity-atoms#194]
**Is your feature request related to a problem? Please describe.**
The Replay Buffer on Events is not always a wanted feature. It can get very tedious to disable it manually on every new event and to make sure no event has it on "by accident" which can lead to sometimes hard to recognize bugs.

**Describe the solution you'd like**
It would be great if there was an option to disable the Replay Buffer on Events. Either by giving the event.Register(Action) method an optional flag to disable the replay buffer for this listener or by generally disabling it on an event asset.

**Describe alternatives you've considered**
Alternatively an option to set a global default value for the Buffer would be an acceptable solution though the safer option would be to have freedom over which listeners want to use the buffer and which do not want to use the buffer.

**Comment by TowerCrow** [comment]
Disabling this feature is the first thing I have to do on every new event. Something that is optional and unnecessary most of the time should not be a setting that's enabled by default.

**Comment by AdamRamberg** [comment]
Thanks for reporting this issue! 

The reason for the Replay Buffer having an initial value of 1 is because a timing issue that will occur if the Replay Buffer is set to 0. Since 1) Variables are triggering there "initial value events" on OnEnable and 2) Listeners are subscribing to Events on OnEnable, it is not certain that Listener's OnEnable will run before Variable's OnEnable. This results in that sometimes Listeners' handlers will run and sometimes not. The current solution (which might not always be preferable) has been the least complex solution to this problem. 

One solution to the issue that I'm describing above might be to use the attribute `DefaultExecutionOrder` for Listeners and Variables. I'm guessing that there is some issues regarding this, since it wasn't done from the beginning, but it's never the less worth investigating again. 

If we can solve the issue that I'm describing above in an elegant way, then I'm fine with the Replay Buffer being initialized to 0. However, I believe that in that case it would be preferable to be able to globally configure Atoms to use another default Replay Buffer value (using user prefs). 

@tivomk Can you please explain your desired solution further? I don't see how your solution is less tedious then manually setting the Replay Buffer to 0 for each newly created Event. Setting the Replay Buffer to 0 on an Event is the same thing as disabling the Replay Buffer for that Event.

**Comment by tivomk** [comment]
The resaon why I would like the solution in code is, that this would ensure the behaviour no matter what is set in the inspector for that particular listener.

So if I say my listener(s) should not receive Replay Buffer calls, then i would love the option to only register for the normal events and not for the buffered ones by default. In terms of tediousness this is of course similar, but the difference being, no one can accidentally fiddle with it or forget to set it to 0 in the event object because it would be ignored on the listeners side then.

The default value settings would be more of a workaround or for projects where it is generally not wanted to use the buffer, but if your described problem doesnt have a good solution then its understandable that this setting isnt gonna be easy to implement.

So to sum it up, the preferred solution would be the option to simply **not** register for the buffer events in code, if you dont want to receive them with that listener when using the .Register method

**Comment by TowerCrow** [comment]
My issue is not that the event buffer exists or has a value > 0. My issue is that "Replay Event Buffer on Register" is enabled by default - this causes events to trigger when they shouldn't and I disable this option 100% of the time I create a new event so it doesn't trigger as soon as I enter Play mode. I don't see why this would be desired as a setting that's enabled by default. It's rather optional.

**Comment by AdamRamberg** [comment]
@tivomk So if I understand you correctly, you want to add the same `bool replayEventsBuffer` as `RegisterListener` has, but to the `Register` function right? I agree that it is indeed strange that we don't have that parameter for the `Register` when we have it for `RegisterListener`. Coming back to this though, I'm not sure that Listeners at all should be concerned about replay buffers, that is really up to the Event, but this is something that we need to look into in the next major of Unity Atoms. 

@TowerCrow Alright! We could solve this by adding an option for overriding default values, via Unity options or in a special / singelton SO. @soraphis @miikalo what are your thoughts on this?

**Comment by soraphis** [comment]
I think 
- default settings should definitely be a thing
- The script execution order problem should definitely be addressed. relying on the replay buffer feels imho more of an hack as the the unity script execution order would. 
- I, too, think replay buffer is part of the event, not the listener.

**Comment by miikalo** [comment]
Added to Backlog milestone as it's not certain this will be in v5.
